partiql-spi/src/main/java/org/partiql/spi/value/Variant.java

```diff
@@ -0,0 +1,43 @@
+package org.partiql.spi.value;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.nio.charset.Charset;
+
+/**
+ * Variant is a {@link Datum} with the ability to pack and unpack a value.
+ *
+ * @param <T>
+ */
+public interface Variant<T> extends Datum {
+
+    /**
+     * Unpack the inner variant value.
+     *
+     * @return T
+     */
+    T unpack();
```

Comment:
It was my assumption that PType.Kind would have another variant: `VARIANT` with some method to expose the parameterization. And, a `Datum` holding type `VARIANT` would allow for the invocation of two methods: a new `Datum unpack()` and the existing `byte[] getBytes()`.

Doing it like this, we could still unpack the variant to a PartiQL value (Datum), and _if_ we expose `IonVariant` as a public class, then Ion's API of `IonElement` wouldn't be exposed either. I _think_ it would provide enough APIs for variant consumers (ourselves included) to create a DatumReader/Writer for the variant type. Also, it would allow for us to internalize `Variant`, `IonDatumReader`, `IonDatumWriter`, and `IonVariant`.

Reply:
Let me know if this clarifies the model.

* A variant _value_ is a datum (partiql _value_) but holds an inner type (hence unpack)
* You do not unpack the variant to a value because it already is a value – variant extends datum
* The VARIANT PType is parameterized by a _string_ aka `variant("ion"), variant("json"), variant("cbor")`
* Unpack is to provide direct access (like old .ionValue)
* Pack is to encode as bytes for the writer.

Hopefully this model is a bit more clear? To address the concerns

* no need to unpack, a _variant_ is a _datum_ implementation
* DatumReader/DatumWriter are the interfaces to read some encoding of PartiQL values into the Datum interface
* Variant, IonDatumReader, IonDatumWriter, and IonVariant are explicitly public.

Reply:
I see you are describing adding variant to the datum fat interface, I thought about that but variant is parameterized on some type. I'm not sure of a better solution without parameterizing Datum.

Reply:
> no need to unpack, a variant is a datum implementation

Yes, exactly. We've established that variant is a runtime type and should be reflected in our runtime values.

In the `IonVariant` impl, there is a TODO to return `PType.variant("ion")` on `getType()`. That's where it's a bit odd to me, because now, we've broken the contract between `getType` and any of the other getters. It seems like the contract would be better between `PType.Kind.VARIANT` and `Datum.unpack(): Datum`.

We're back to a similar `.check<Variant>()` as before, where we can't fully trust the relationship between `Datum#getType()` and the getters. For example:
```kotlin
// This is all pseudo-code, for brevity
class Udf : Function {
    override fun signature() = Signature("udf", params = listOf(PType.variantIon()))

    override fun eval(args: Array<Datum>) {
        val datum = args[0]
        val variant: IonVariant = datum.check<IonVariant>() // which will throw a TypeCheckException and perform the cast
        val underlyingValue: IonElement = datum.unpack()
        return // whatever we want
    }
}
```

Rather than:
```kotlin
// This is all pseudo-code, for brevity
class Udf : Function {
    override fun signature() = Signature("udf", params = listOf(PType.variantIon()))

    override fun eval(args: Array<Datum>) {
        val underlyingValue: Datum = datum.unpack() // Or, we can call datum.getBytes()
        return // whatever we want
    }
}
```

This is especially apparent when we want to be in charge of casting. How would we know which Datum getter to invoke when `getType()` returns `PType.variant("ion")`?